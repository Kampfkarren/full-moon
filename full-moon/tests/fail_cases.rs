use full_moon::{
    ast,
    tokenizer::{self, Token},
};
use insta::assert_yaml_snapshot;
use pretty_assertions::assert_eq;
use std::fs::{self, File};
use std::io::Write;

mod common;
use common::run_test_folder;

#[test]
#[cfg_attr(feature = "no-source-tests", ignore)]
fn test_parser_fail_cases() {
    run_test_folder("./tests/cases/fail/parser", |path| {
        let source = fs::read_to_string(path.join("source.lua")).expect("couldn't read source.lua");

        let tokens = tokenizer::tokens(&source).expect("couldn't tokenize");

        let tokens_path = path.join("tokens.json");
        let tokens_contents;

        if let Ok(tokens_contents_tmp) = fs::read_to_string(dbg!(&tokens_path)) {
            tokens_contents = tokens_contents_tmp;
            let expected_tokens: Vec<Token> =
                serde_json::from_str(&tokens_contents).expect("couldn't deserialize tokens file");
            {
                let tokens = &expected_tokens;
                assert_yaml_snapshot!("tokens", tokens);
            }
            assert_eq!(tokens, expected_tokens);
        } else {
            let mut file = File::create(&tokens_path).expect("couldn't write tokens file");
            file.write_all(
                serde_json::to_string_pretty(&tokens)
                    .expect("couldn't serialize")
                    .as_bytes(),
            )
            .expect("couldn't write to tokens file");
        }

        match ast::Ast::from_tokens(tokens) {
            Ok(_) => panic!("fail case passed for {:?}", path),
            Err(error) => {
                println!("error {:#?}", error);
                let error_path = path.join("error.json");
                if let Ok(error_contents) = fs::read_to_string(&error_path) {
                    let expected_error = serde_json::from_str(&error_contents)
                        .expect("couldn't deserialize existing error file");
                    {
                        let error = &expected_error;
                        assert_yaml_snapshot!("error", error);
                    }
                    assert_eq!(error, expected_error);
                } else {
                    let mut file = File::create(&error_path).expect("couldn't write error file");
                    file.write_all(
                        serde_json::to_string_pretty(&error)
                            .expect("couldn't serialize")
                            .as_bytes(),
                    )
                    .expect("couldn't write to ast file");
                }
            }
        }
    })
}

#[test]
#[cfg_attr(feature = "no-source-tests", ignore)]
fn test_tokenizer_fail_cases() {
    run_test_folder("./tests/cases/fail/tokenizer", |path| {
        let source = fs::read_to_string(path.join("source.lua")).expect("couldn't read source.lua");

        match tokenizer::tokens(&source) {
            Ok(_) => panic!("fail case passed for {:?}", path),
            Err(error) => {
                let error_path = path.join("error.json");
                if let Ok(error_contents) = fs::read_to_string(&error_path) {
                    let expected_error = serde_json::from_str(&error_contents)
                        .expect("couldn't deserialize existing error file");
                    {
                        let error = &expected_error;
                        assert_yaml_snapshot!("error", error);
                    }
                    assert_eq!(error, expected_error);
                } else {
                    let mut file = File::create(&error_path).expect("couldn't write error file");
                    file.write_all(
                        serde_json::to_string_pretty(&error)
                            .expect("couldn't serialize")
                            .as_bytes(),
                    )
                    .expect("couldn't write to ast file");
                }
            }
        }
    })
}

#[test]
#[cfg(feature = "lua52")]
#[cfg_attr(feature = "no-source-tests", ignore)]
fn test_lua52_parser_fail_cases() {
    run_test_folder("./tests/lua52_cases/fail/parser", |path| {
        let source = fs::read_to_string(path.join("source.lua")).expect("couldn't read source.lua");

        let tokens = tokenizer::tokens(&source).expect("couldn't tokenize");

        let tokens_path = path.join("tokens.json");
        let tokens_contents;

        if let Ok(tokens_contents_tmp) = fs::read_to_string(dbg!(&tokens_path)) {
            tokens_contents = tokens_contents_tmp;
            let expected_tokens: Vec<Token> =
                serde_json::from_str(&tokens_contents).expect("couldn't deserialize tokens file");
            {
                let tokens = &expected_tokens;
                assert_yaml_snapshot!("tokens", tokens);
            }
            assert_eq!(tokens, expected_tokens);
        } else {
            let mut file = File::create(&tokens_path).expect("couldn't write tokens file");
            file.write_all(
                serde_json::to_string_pretty(&tokens)
                    .expect("couldn't serialize")
                    .as_bytes(),
            )
            .expect("couldn't write to tokens file");
        }

        match ast::Ast::from_tokens(tokens) {
            Ok(_) => panic!("fail case passed for {:?}", path),
            Err(error) => {
                println!("error {:#?}", error);
                let error_path = path.join("error.json");
                if let Ok(error_contents) = fs::read_to_string(&error_path) {
                    let expected_error = serde_json::from_str(&error_contents)
                        .expect("couldn't deserialize existing error file");
                    {
                        let error = &expected_error;
                        assert_yaml_snapshot!("error", error);
                    }
                    assert_eq!(error, expected_error);
                } else {
                    let mut file = File::create(&error_path).expect("couldn't write error file");
                    file.write_all(
                        serde_json::to_string_pretty(&error)
                            .expect("couldn't serialize")
                            .as_bytes(),
                    )
                    .expect("couldn't write to ast file");
                }
            }
        }
    })
}
