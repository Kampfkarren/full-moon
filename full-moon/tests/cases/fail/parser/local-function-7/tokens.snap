---
source: full-moon/tests/fail_cases.rs
assertion_line: 28
expression: tokens
input_file: full-moon/tests/cases/fail/parser/local-function-7
---
- start_position:
    bytes: 0
    line: 1
    character: 1
  end_position:
    bytes: 5
    line: 1
    character: 6
  token_type:
    type: Symbol
    symbol: local
- start_position:
    bytes: 5
    line: 1
    character: 6
  end_position:
    bytes: 6
    line: 1
    character: 7
  token_type:
    type: Whitespace
    characters: " "
- start_position:
    bytes: 6
    line: 1
    character: 7
  end_position:
    bytes: 14
    line: 1
    character: 15
  token_type:
    type: Symbol
    symbol: function
- start_position:
    bytes: 14
    line: 1
    character: 15
  end_position:
    bytes: 15
    line: 1
    character: 16
  token_type:
    type: Whitespace
    characters: " "
- start_position:
    bytes: 15
    line: 1
    character: 16
  end_position:
    bytes: 18
    line: 1
    character: 19
  token_type:
    type: Identifier
    identifier: foo
- start_position:
    bytes: 18
    line: 1
    character: 19
  end_position:
    bytes: 19
    line: 1
    character: 20
  token_type:
    type: Symbol
    symbol: (
- start_position:
    bytes: 19
    line: 1
    character: 20
  end_position:
    bytes: 20
    line: 1
    character: 21
  token_type:
    type: Identifier
    identifier: x
- start_position:
    bytes: 20
    line: 1
    character: 21
  end_position:
    bytes: 21
    line: 1
    character: 22
  token_type:
    type: Symbol
    symbol: ","
- start_position:
    bytes: 21
    line: 1
    character: 22
  end_position:
    bytes: 22
    line: 1
    character: 23
  token_type:
    type: Whitespace
    characters: " "
- start_position:
    bytes: 22
    line: 1
    character: 23
  end_position:
    bytes: 23
    line: 1
    character: 24
  token_type:
    type: Identifier
    identifier: y
- start_position:
    bytes: 23
    line: 1
    character: 24
  end_position:
    bytes: 24
    line: 1
    character: 25
  token_type:
    type: Symbol
    symbol: ","
- start_position:
    bytes: 24
    line: 1
    character: 25
  end_position:
    bytes: 25
    line: 1
    character: 26
  token_type:
    type: Symbol
    symbol: )
- start_position:
    bytes: 25
    line: 1
    character: 26
  end_position:
    bytes: 26
    line: 1
    character: 26
  token_type:
    type: Whitespace
    characters: "\n"
- start_position:
    bytes: 26
    line: 2
    character: 1
  end_position:
    bytes: 27
    line: 2
    character: 2
  token_type:
    type: Whitespace
    characters: "\t"
- start_position:
    bytes: 27
    line: 2
    character: 2
  end_position:
    bytes: 32
    line: 2
    character: 7
  token_type:
    type: Identifier
    identifier: print
- start_position:
    bytes: 32
    line: 2
    character: 7
  end_position:
    bytes: 33
    line: 2
    character: 8
  token_type:
    type: Symbol
    symbol: (
- start_position:
    bytes: 33
    line: 2
    character: 8
  end_position:
    bytes: 34
    line: 2
    character: 9
  token_type:
    type: Identifier
    identifier: x
- start_position:
    bytes: 34
    line: 2
    character: 9
  end_position:
    bytes: 35
    line: 2
    character: 10
  token_type:
    type: Symbol
    symbol: ","
- start_position:
    bytes: 35
    line: 2
    character: 10
  end_position:
    bytes: 36
    line: 2
    character: 11
  token_type:
    type: Whitespace
    characters: " "
- start_position:
    bytes: 36
    line: 2
    character: 11
  end_position:
    bytes: 37
    line: 2
    character: 12
  token_type:
    type: Identifier
    identifier: y
- start_position:
    bytes: 37
    line: 2
    character: 12
  end_position:
    bytes: 38
    line: 2
    character: 13
  token_type:
    type: Symbol
    symbol: )
- start_position:
    bytes: 38
    line: 2
    character: 13
  end_position:
    bytes: 39
    line: 2
    character: 13
  token_type:
    type: Whitespace
    characters: "\n"
- start_position:
    bytes: 39
    line: 3
    character: 1
  end_position:
    bytes: 42
    line: 3
    character: 4
  token_type:
    type: Symbol
    symbol: end
- start_position:
    bytes: 42
    line: 3
    character: 4
  end_position:
    bytes: 43
    line: 3
    character: 4
  token_type:
    type: Whitespace
    characters: "\n"
- start_position:
    bytes: 43
    line: 4
    character: 1
  end_position:
    bytes: 43
    line: 4
    character: 1
  token_type:
    type: Eof

